library/ecdh.c

```diff
@@ -726,4 +727,88 @@ int mbedtls_ecdh_calc_secret( mbedtls_ecdh_context *ctx, size_t *olen,
 #endif
 }
 
+#if defined(MBEDTLS_SSL_PROTO_TLS1_3_EXPERIMENTAL)
+
+static int ecdh_tls13_make_params_internal( mbedtls_ecdh_context_mbed *ctx,
+                                            size_t *olen, int point_format,
+                                            unsigned char *buf, size_t blen,
+                                            int ( *f_rng )( void *,
+                                                            unsigned char *,
+                                                            size_t),
+                                            void *p_rng, int restart_enabled )
+{
+    int ret = MBEDTLS_ERR_ERROR_CORRUPTION_DETECTED;
+#if defined(MBEDTLS_ECP_RESTARTABLE)
+    mbedtls_ecp_restart_ctx *rs_ctx = NULL;
+#endif
+
+    if( ctx->grp.pbits == 0 )
+        return( MBEDTLS_ERR_ECP_BAD_INPUT_DATA );
+
+#if defined(MBEDTLS_ECP_RESTARTABLE)
+    if( restart_enabled )
+        rs_ctx = &ctx->rs;
+#else
+    (void) restart_enabled;
+#endif
+
+#if defined(MBEDTLS_ECP_RESTARTABLE)
+    if( ( ret = ecdh_gen_public_restartable( &ctx->grp, &ctx->d, &ctx->Q,
+                                             f_rng, p_rng, rs_ctx ) ) != 0 )
+        return( ret );
+#else
+    if( ( ret = mbedtls_ecdh_gen_public( &ctx->grp, &ctx->d, &ctx->Q,
```

Comment:
Need Help

Reply:
I think that might be part of #4824 .

Reply:
It is indeed.

Reply:
Looking at this again, it seems that "restartable" is enabled only for TLS 1.2 in ssl_cli.c. I guess that enabling it for TLS 1.3 would mean much more than just calling `mbedtls_ecdh_tls13_make_params()` here thus calling ECDH code that is not "restartable" capable ( i.e mbedtls_ecdh_gen_public() and mbedtls_ecp_point_write_binary() ) shouldn't be a problem for the time being.

Reply:
First thing I want to say is, wow this area is messy. Between restartable, everest, legacy context, alt implementations, and the legacy context having members that aren't used anywhere in the code (they were added with future improvements in mind)... I suggest keeping things simple for now in the interest of advancing quickly: no support for restartable, no support for everest.

Regarding restartable, as Ronald said, it wouldn't be just here, and also it would need to be tested; I don't think that would be very hard but still a non-zero amount of work. IMO that's not a necessary part of an MVP, and can be added later. This part is easy: just call the functions that don't have `_restartable` in the name and that's it.

Regarding Everest, again as Ronald said, I suspect support is currently broken; it would have to be tested, and getting it to work would also be complex as the Everest API was unfortunately tied to the TLS 1.2 encoding, so we'd had to transcode, to/from TLS 1.3 encoding which can be done but would be tedious. Also, the same point apply: we're doing an MVP, support for Everest is an optimisation that can be done later. (Also, these days, Everest should really be a PSA driver rather than use an ad-hoc API, so perhaps we don't even want to support it in 1.3 in its current form.)

This time implementation is a bit less clear though, as unlike restartable which has to be selected explicitly by calling a specific set of functions Everest is used automatically for Curve25519 as long as support is compile in. So, I can see two options here:

1. Refuse to build with `MBEDTLS_SSL_PROTO_TLS1_3_EXPERIMENTAL` and `MBEDTLS_ECDH_VARIANT_EVEREST_ENABLED` at the same time. This doesn't feel right, people might want to play with 1.3 (without Everest) _and_ still use Everest in 1.2 in the same build.
2. Introduce a new private function `mbedtls_ecdh_setup_no_everest()` that sets up a context without everest (basically the same as `mbedtls_ecdh_setup()` but without the `switch` on the `grp_id`) and use that in the TLS 1.3 code. No extremely clean as it means some duplication, but that's purely internal and can be cleaned up later - we have a lot of cleaning to do in this area anyway. I think that's my favourite option unless one of you can suggest something better.

Then the only point remaining is about `LEGACY_CONTEXT`. Here I think the current code structure is fine: have `ecdh_tls13_make_params_internal()` (that would only call `mbedtls_ecdh_gen_public()` and `mbedtls_ecp_point_write_binary()` once it's been stripped from restartable and everest support) and then `mbedtls_ecdh_tls13_make_params()` calls that, passing either `ctx` or `ctx->ctx.mbed_ecdh` to it depending of whether `LEGACY_CONTEXT` is enabled or not. (Note: this is, I think, the only point where I disagree with Ronald: IMO we should keep those functions instead of calling `mbedtls_ecdh_gen_public()` and `mbedtls_ecp_point_write_binary()` directly from `ssl_tls13_client.c`.)

Also, now that I think of it, we need to test multiple builds. At least one with Everest enabled (and restartable disabled, as these two are incompatible) and one with restartable enabled (hence everest disabled), perhaps also one with both disabled (default config). Perhaps this is already covered by existing `all.sh` components, but we should double-check. (Note: legacy context is now tied to restartable: it's enabled exactly when restartable is.)

Reply:
`MBEDTLS_ECDH_LEGACY_CONTEXT` is no longer a config option. In 3.x, it's only enabled when restartable ECC is enabled. So there are only three cases: Everest, portable C (with the new context), and restartable.

Reply:
It's not a independent config option, but I think we still need to make a distinction between:

1. supporting builds that have `MBEDTLS_ECP_RESTARTABLE` (hence `MBEDTLS_ECDH_LEGACY_CONTEXT`) enabled, which I think we want to do even for the MVP
2. supporting actual use of the restartable feature with TLS 1.3, which I think we should declare out of scope of the MVP.

I agree there are only three cases, that's the three build options I suggested for tests (now that I think of it, we need to test the three of them, the "perhaps" was wrong). But I think in all of the three builds we should do the same thing _for the MVP_: use our own implementation (not everest) without support for restarting operations.

Reply:
There are two independent issues here:
1. Whether to worry about Everest and restartable ECC for the TLS 1.3 MVP
2. How to best organize the ECDHE code, accounting for the slightly different encoding of ECC related structures in TLS 1.3

I would like to request that we agree and fix 1 here, but not 2, which is non-trivial and the subject of issue #4824. It's IMO acceptable to have potential duplication between the TLS 1.2 and TLS 1.3 functions until we've analyzed precisely which bits can be shared and which do require separate code for TLS 1.3.

Reply:
I add two new functions into #4824 , they should be same issue.

Reply:
What's action I should take ? 

Remove everest and restartable from TLS 1.3 code ?